### PR TITLE
Fix build and begin some refactor+testing

### PR DESCRIPTION
**Pull request by saxon-milton**
This PR encompasses changes that firstly fix build and remove dependencies on repos that are private. There are is some refactoring of parsing functionality for syntax elements with descriptors ue, se, te, and me. The file bit_reader.go has been replaced by read.go, where reading utilities will temporarily stay.  (NB the deltas in read.go are just what has been transferred from other files, or the deleted bit_reader.go file, not sure why it didn't register as a file rename).

### CODE REVIEW

File: h264/parse.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
+/*
```
Comment (saxon-milton) [inline h264/parse.go:1]: Need to fix this file header.

File: h264/parse_test.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
+/*
```
Comment (saxon-milton) [inline h264/parse_test.go]: Need to fix this file header

**Comment by scruzin** [comment]
Curious why the various xy() functions are now readXy()?

File: h264/parse.go [hunk 3]
```diff
@@ -93,3 +93,69 @@ func readSe(r bitio.Reader) (int, error) {
...
+	case inter:
+		i3 = 1
+	default:
+		return 0, errors.New("invalid macroblock prediction mode")
+	}
+
+	return codedBlockPattern[i1][i2][i3], nil
+}
+
+// TODO: this should probably go in a separate consts file like consts.go or something
```
Comment (scruzin) [inline h264/parse.go]: Yes, please refactor into a separate file, or at the very least move to the top of the file.

File: h264/parse.go [hunk 4]
```diff
@@ -108,7 +108,7 @@ func readMe(r bitio.Reader, chromaArrayType int, mpm macroblockPredictionMode) (
 	case 0, 3:
 		i1 = 1
 	default:
-		return 0, errors.New("invalid chromaArrayType")
+		return 0, errInvalidCAT
```
Comment (scruzin) [inline h264/parse.go]: Are these errors every likely to be exposed outside of the package, i.e., to callers?

File: h264/parse.go [hunk 5]
```diff
@@ -117,19 +117,31 @@ func readMe(r bitio.Reader, chromaArrayType int, mpm macroblockPredictionMode) (
...
 		i3 = 1
 	default:
-		return 0, errors.New("invalid macroblock prediction mode")
+		return 0, errInvalidMPM
 	}
 
 	return codedBlockPattern[i1][i2][i3], nil
 }
 
+// Errors used by readMe.
```
Comment (scruzin) [inline h264/parse.go:131]: Are these errors every likely to be exposed outside of the package, i.e., to callers?

File: h264/parse.go [hunk 6]
```diff
@@ -117,19 +117,31 @@ func readMe(r bitio.Reader, chromaArrayType int, mpm macroblockPredictionMode) (
...
+var (
+	errInvalidCodeNum = errors.New("invalid codeNum")
```
Comment (scruzin) [inline h264/parse.go:133]: invalid -> Invalid

**Comment by scruzin** [comment]
LGTM from a style standpoint. Can't really comment otherwise.

**Comment by ausocean** [comment]
LGTM from a style standpoint. Can't really comment otherwise.

**Comment by scruzin** [comment]
LGTM from a style standpoint. Can't really comment otherwise.

**Review by scruzin — APPROVED** [review]
LGTM from a style standpoint. Can't really comment otherwise.

**Comment by saxon-milton** [comment]
> Curious why the various xy() functions are now readXy()?

The xy names are are descriptor types for syntax elements.  It seemed clearer to me to call the functions that read and parse the syntax elements readXy, but I could be wrong.

File: h264/parse.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+const (
+	intra4x4 macroblockPredictionMode = iota
+	intra8x8
+	inter
+)
+
+// codedBlockPattern contains data from table 9-4 in ITU-T H.264 (04/2017)
+// for mapping a chromaArrayType, codeNum and macroblock prediction mode to a
+// coded block pattern.
+var codedBlockPattern = [][][]uint{
```
Comment (kortschak) [inline h264/parse.go]: `[][][2]uint`

File: h264/parse_test.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+		{[]byte{0x28}, 4},  // Bit string: 00101, codeNum: 4
+		{[]byte{0x30}, 5},  // Bit string: 00110, codeNum: 5
+		{[]byte{0x38}, 6},  // Bit string: 00111, codeNum: 6
+		{[]byte{0x10}, 7},  // Bit string: 0001000, codeNum: 7
+		{[]byte{0x12}, 8},  // Bit string: 0001001, codeNum: 8
+		{[]byte{0x14}, 9},  // Bit string: 0001010, codeNum: 9
+		{[]byte{0x16}, 10}, // Bit string: 0001011, codeNum: 10
+	}
+
+	for testn, test := range tests {
```
Comment (kortschak) [inline h264/parse_test.go]: s/testn/i/g

File: h264/parse_test.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+)
+
+// TestReadUe checks that readUe correctly parses an Exp-Golomb-coded element
+// to a code number.
+func TestReadUe(t *testing.T) {
+	// tests has been derived from Table 9-2 in ITU-T H.H264, showing bit strings
+	// and corresponding codeNums.
+	tests := []struct {
+		in     []byte // The bitstring we wish to read.
+		expect uint   // The expected codeNum.
```
Comment (kortschak) [inline h264/parse_test.go]: s/expect/want/g

File: h264/parse_test.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+		{[]byte{0x16}, 10}, // Bit string: 0001011, codeNum: 10
+	}
+
+	for testn, test := range tests {
+		got, err := readUe(bitio.NewReader(bytes.NewReader(test.in)))
+		if err != nil {
+			t.Fatalf("did not expect to get error: %v from readUe", err)
+		}
+
+		if test.expect != uint(got) {
```
Comment (kortschak) [inline h264/parse_test.go]: Reverse order: "I didn't get what I wanted".

File: h264/pps.go [hunk 11]
```diff
@@ -49,27 +49,27 @@ func NewPPS(sps *SPS, rbsp []byte, showPacket bool) *PPS {
 		return false
 	}
 
-	pps.ID = ue(b.golomb())
-	pps.SPSID = ue(b.golomb())
+	pps.ID, _ = readUe(nil)
```
Comment (kortschak) [inline h264/pps.go]: Why are the errors dropped?

File: h264/read.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
+package h264
```
Comment (kortschak) [inline h264/read.go:1]: Header.

File: h264/read.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+type BitReader struct {
```
Comment (kortschak) [inline h264/read.go:9]: Comments?

File: h264/read.go [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+
+// TODO: what does this do ?
+func bitVal(bits []int) int {
+	t := 0
+	for i, b := range bits {
+		if b == 1 {
+			t += 1 << uint((len(bits)-1)-i)
+		}
+	}
+	// fmt.Printf("\t bitVal: %d\n", t)
```
Comment (kortschak) [inline h264/read.go:57]: Delete this?

File: h264/read.go [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+func (r *H264Reader) readNalUnit() (*NalUnit, *BitReader) {
+	// Read to start of NAL
+	logger.Printf("debug: Seeking NAL %d start\n", len(r.NalUnits))
+	r.LogStreamPosition()
+	for !isStartSequence(r.Bytes()) {
+		if err := r.BufferToReader(1); err != nil {
+			return nil, nil
+		}
+	}
+	/*
```
Comment (kortschak) [inline h264/read.go:97]: ?

**Comment by shawnps** [comment]
LGTM - @kortschak made a lot of nice suggestions.

File: h264/pps.go (hunk 11 above)
Comment (saxon-milton) [reply h264/pps.go]: This is intended to be temporary. Should I do add a TODO, or actually handle the errors now ?

File: h264/read.go (hunk 13 above)
Comment (saxon-milton) [reply h264/read.go:9]: None of the code in this file is my code. This should have been a file rename - I must have done something weird.

File: h264/parse.go (hunk 6 above)
Comment (saxon-milton) [reply h264/parse.go:133]: I was under the impression go error messages shouldn't be capitalised

Comment (shawnps) [reply h264/parse.go:133]: That is my impression as well: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: h264/pps.go (hunk 11 above)
Comment (kortschak) [reply h264/pps.go]: The errors should be handled now.

File: h264/parse.go (hunk 5 above)
Comment (saxon-milton) [reply h264/parse.go:131]: I'm not sure at this point if they will be exported.

**Pull request merged**